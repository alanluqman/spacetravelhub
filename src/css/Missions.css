* {
  box-sizing: border-box;
}

.container {
  max-width: 1000px;
  margin: auto;
  margin-top: 20px;
  padding: 10px;
}

.table {
  width: 100%;
}

table,
th,
td {
  border: 1px solid #868686;
  border-collapse: collapse;
  text-align: left;
  padding: 3px;
}

tr:nth-child(even) {
  background-color: #d3d3d3;
}
