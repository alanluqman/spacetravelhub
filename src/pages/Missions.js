/* eslint-disable max-len */
import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import axios from 'axios';
import { fetchMissions } from '../redux/missions';
import Mission from '../component/mission';
import '../css/Missions.css';

export default function Missions() {
  const missionList = useSelector((state) => state.mission);
  const dispatch = useDispatch();
  const api = 'https://api.spacexdata.com/v3/missions';
  useEffect(() => async () => {
    const data = await axios.get(api).then((response) => response.data);
    // eslint-disable-next-line prefer-const
    let newList = [];
    data.forEach((element) => {
      newList.push({
        mission_id: element.mission_id,
        mission_name: element.mission_name,
        description: element.description,
        reserved: false,
      });
    });
    dispatch(fetchMissions(newList));
  }, []);
  return (
    <div className="container">
      <table className="table">
        <thead>
          <tr>
            <th style={{ width: '20%' }}>Mission</th>
            <th style={{ width: '50%' }}>Description</th>
            <th style={{ width: '15%' }}>Status</th>
            <th style={{ width: '15%' }}> </th>
          </tr>
        </thead>
        <tbody>
          {missionList.map((mission) => <Mission key={mission.mission_id} id={mission.mission_id} name={mission.mission_name} description={mission.description} reserved={mission.reserved} />)}
        </tbody>
      </table>
    </div>
  );
}
