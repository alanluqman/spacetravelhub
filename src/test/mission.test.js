import {
  render, screen, act, fireEvent,
} from '@testing-library/react';
import { Provider } from 'react-redux';
import store from '../redux/configureStore';
import Missions from '../pages/Missions';

describe('Missions page component', () => {
  afterEach(() => {
    act(() => store.dispatch({
      type: 'spacetravelhub/missions/fetch',
      payload: [],
    }));
  });
  // check for missions table
  it('Check if table of missions available', async () => {
    render(
      <Provider store={store}>
        <Missions />
      </Provider>,
    );
    expect(screen.getByText('Description')).toMatchSnapshot();
  });
  // check for join mission buttons
  it('Should join mission button available', async () => {
    render(
      <Provider store={store}>
        <Missions />
      </Provider>,
    );
    const button = await screen.findAllByText(/Join Mission/i);
    expect(button).toMatchSnapshot();
  });
  // check for NOT A MEMBER badge
  it('Should NOT A MEMBER badge available', async () => {
    render(
      <Provider store={store}>
        <Missions />
      </Provider>,
    );
    const notMemberBadge = await screen.findAllByText(/NOT A MEMBER/i);
    expect(notMemberBadge).toMatchSnapshot();
  });
  // check for Leave Mission button if i clicked Join Mission
  it('Should Leave Mission button available', async () => {
    render(
      <Provider store={store}>
        <Missions />
      </Provider>,
    );
    const button = await screen.findAllByText('Join Mission');
    fireEvent.click(button[0]);
    const leaveMissionBtn = await screen.findAllByText(/Leave Mission/i);
    expect(leaveMissionBtn).toMatchSnapshot();
  });
  // check for Active Member badge when i clicked Join Mission button
  it('Should Active Member badge available', async () => {
    render(
      <Provider store={store}>
        <Missions />
      </Provider>,
    );
    const button = await screen.findAllByText('Join Mission');
    fireEvent.click(button[0]);
    const activeMemberBadge = await screen.findAllByText(/Active Member/i);
    expect(activeMemberBadge).toMatchSnapshot();
  });
});
