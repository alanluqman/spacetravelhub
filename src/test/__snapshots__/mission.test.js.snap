// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Missions page component Check if table of missions available 1`] = `
<th
  style="width: 50%;"
>
  Description
</th>
`;

exports[`Missions page component Should Active Member badge available 1`] = `
Array [
  <span
    style="background-color: rgb(0, 163, 136); color: white; padding: 5px; border-radius: 5px;"
  >
    Active Member
  </span>,
]
`;

exports[`Missions page component Should Leave Mission button available 1`] = `
Array [
  <button
    style="background-color: transparent; border: 1px solid; border-color: #f00; color: rgb(255, 0, 0); padding: 5px; border-radius: 5px; cursor: pointer;"
    type="button"
  >
    Leave Mission
  </button>,
]
`;

exports[`Missions page component Should NOT A MEMBER badge available 1`] = `
Array [
  <span
    style="background-color: rgb(83, 83, 83); color: white; padding: 5px; border-radius: 5px;"
  >
    NOT A MEMBER
  </span>,
  <span
    style="background-color: rgb(83, 83, 83); color: white; padding: 5px; border-radius: 5px;"
  >
    NOT A MEMBER
  </span>,
  <span
    style="background-color: rgb(83, 83, 83); color: white; padding: 5px; border-radius: 5px;"
  >
    NOT A MEMBER
  </span>,
  <span
    style="background-color: rgb(83, 83, 83); color: white; padding: 5px; border-radius: 5px;"
  >
    NOT A MEMBER
  </span>,
  <span
    style="background-color: rgb(83, 83, 83); color: white; padding: 5px; border-radius: 5px;"
  >
    NOT A MEMBER
  </span>,
  <span
    style="background-color: rgb(83, 83, 83); color: white; padding: 5px; border-radius: 5px;"
  >
    NOT A MEMBER
  </span>,
  <span
    style="background-color: rgb(83, 83, 83); color: white; padding: 5px; border-radius: 5px;"
  >
    NOT A MEMBER
  </span>,
  <span
    style="background-color: rgb(83, 83, 83); color: white; padding: 5px; border-radius: 5px;"
  >
    NOT A MEMBER
  </span>,
  <span
    style="background-color: rgb(83, 83, 83); color: white; padding: 5px; border-radius: 5px;"
  >
    NOT A MEMBER
  </span>,
  <span
    style="background-color: rgb(83, 83, 83); color: white; padding: 5px; border-radius: 5px;"
  >
    NOT A MEMBER
  </span>,
]
`;

exports[`Missions page component Should join mission button available 1`] = `
Array [
  <button
    style="background-color: transparent; border: 1px solid; border-color: #535353; color: rgb(83, 83, 83); padding: 5px; border-radius: 5px; cursor: pointer;"
    type="button"
  >
    Join Mission
  </button>,
  <button
    style="background-color: transparent; border: 1px solid; border-color: #535353; color: rgb(83, 83, 83); padding: 5px; border-radius: 5px; cursor: pointer;"
    type="button"
  >
    Join Mission
  </button>,
  <button
    style="background-color: transparent; border: 1px solid; border-color: #535353; color: rgb(83, 83, 83); padding: 5px; border-radius: 5px; cursor: pointer;"
    type="button"
  >
    Join Mission
  </button>,
  <button
    style="background-color: transparent; border: 1px solid; border-color: #535353; color: rgb(83, 83, 83); padding: 5px; border-radius: 5px; cursor: pointer;"
    type="button"
  >
    Join Mission
  </button>,
  <button
    style="background-color: transparent; border: 1px solid; border-color: #535353; color: rgb(83, 83, 83); padding: 5px; border-radius: 5px; cursor: pointer;"
    type="button"
  >
    Join Mission
  </button>,
  <button
    style="background-color: transparent; border: 1px solid; border-color: #535353; color: rgb(83, 83, 83); padding: 5px; border-radius: 5px; cursor: pointer;"
    type="button"
  >
    Join Mission
  </button>,
  <button
    style="background-color: transparent; border: 1px solid; border-color: #535353; color: rgb(83, 83, 83); padding: 5px; border-radius: 5px; cursor: pointer;"
    type="button"
  >
    Join Mission
  </button>,
  <button
    style="background-color: transparent; border: 1px solid; border-color: #535353; color: rgb(83, 83, 83); padding: 5px; border-radius: 5px; cursor: pointer;"
    type="button"
  >
    Join Mission
  </button>,
  <button
    style="background-color: transparent; border: 1px solid; border-color: #535353; color: rgb(83, 83, 83); padding: 5px; border-radius: 5px; cursor: pointer;"
    type="button"
  >
    Join Mission
  </button>,
  <button
    style="background-color: transparent; border: 1px solid; border-color: #535353; color: rgb(83, 83, 83); padding: 5px; border-radius: 5px; cursor: pointer;"
    type="button"
  >
    Join Mission
  </button>,
]
`;
